[STORM-3388] Launch workers inside container using runc runtime

Ethanlm opened the pull request:
## What is the purpose of the change

Enabled supervisors to use runc binary to launch workers inside an oci container. See docs/OCI-support.md for justifications. The original design and many parts of the oci related code is borrowed from Verizon Media hadoop-core team 

This has been used in our production Storm clusters for more than one year. 

## How was the change tested

Tested with an example WordCount topology, including following operations:
1. launch a topology and supervisor launches the worker inside the oci container
```
-bash-4.2$ sudo runc list
ID                                          PID         STATUS      BUNDLE                                                           CREATED                          OWNER
6703-1a23ca4b-6062-4d08-8ac3-b09e7d35e7cb   21780       running     /home/y/var/storm/workers/1a23ca4b-6062-4d08-8ac3-b09e7d35e7cb   2020-12-21T20:21:17.116971511Z   root
```
2. manually kill the worker, supervisor recovers it
3. profile the worker from UI (jmap, heap dump, etc)
```
-bash-4.2$ sudo ls -ltrh /home/y/var/storm/workers-artifacts/wc1-2-1608581491/6703/
...
-rw-r----- 1 username1 gstorm  64K Dec 21 21:32 worker.log
-rw-r----- 1 username1 gstorm  27K Dec 21 21:51 jstack-17-20201221215145.txt
-rw-r----- 1 username1 gstorm 552M Dec 21 21:52 recording-17-20201221215222.bin
```

4. Checked cpu throttling and validated it was properly enforced. In this example, this worker is assigned with 140% cpu.
```
-bash-4.2$ cat /sys/fs/cgroup/cpu/storm/6703-1a23ca4b-6062-4d08-8ac3-b09e7d35e7cb/cpu.cfs_period_us
100000
-bash-4.2$ cat /sys/fs/cgroup/cpu/storm/6703-1a23ca4b-6062-4d08-8ac3-b09e7d35e7cb/cpu.cfs_quota_us
140000
-bash-4.2$ cat /sys/fs/cgroup/cpu/storm/6703-1a23ca4b-6062-4d08-8ac3-b09e7d35e7cb/cpu.stat
nr_periods 824
nr_throttled 821
throttled_time 104646268666
   
  #Top output:
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
21800 username1  20   0 5453928   1.7g  28036 S 139.9 22.7   2:15.96 java
21780 username1  20   0 2451376  83324  27448 S   0.0  1.1   0:02.36 java
```

5. kill the topology, so supervisor kills the worker

Ethanlm commented:
The travis build passed. Squashed the commits

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+      chmod_dirs.append(dir_path)
+  hdfs_chmod("755", chmod_dirs)
+
+def append_or_extend_to_list(src, src_list):
```
Parameter order here is counter-intuitive. It makes more sense if reversed.

Ethanlm replied:
It means appending `src` to `src_list`. I think this order makes sense

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+      exception_string = ("Commmand: " + str(command)
+                          + " failed with returncode: "
+                          + str(process.returncode))
+      if out != None:
```
I believe PEP8 style would be "if out is not None:"

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+
+def shell_command(command, print_stdout, print_stderr, raise_on_error,
+                  timeout_sec=600):
+  global LOG_LEVEL
```
Indentation should be 4 spaces (PEP8)

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+      process.kill()
+    raise Exception("Popen failure")
+  finally:
+    if timer:
```
timer may not be initialized; need a timer = None at line 55

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+
+def get_layer_hashes_from_manifest(manifest, error_on_size_check=True):
+  layers = []
+  size = 0;
```
Remove semicolon

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+  global HADOOP_PREFIX
+  global HADOOP_BIN_DIR
+  global MAX_IMAGE_LAYERS
+  global MAX_IMAGE_SIZE
```
Global var MAX_IMAGE_SIZE is not defined in this module

Ethanlm replied:
They are passed in from arguments

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+    out, err = process.communicate()
+
+    if raise_on_error and process.returncode is not 0:
+      exception_string = ("Commmand: " + str(command)
```
sp: Commmand -> Command

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+    for file in glob.glob(jar_path):
+      jar_file = file
+
+    if not jar_file:
```
replace with"else:" or initialize jar_file = None before for loop

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+
+      elif sub_command == "remove-tag":
+        tags = image_and_tag_arg.split(",")
+        image = None
+        manifest = None
+        manifest_hash = 0
```
Unused - can be removed

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+        if manifest_hash:
+          manifest_path = hdfs_manifest_dir + "/" + manifest_hash
+          out, err, returncode = hdfs_cat(manifest_path)
+          manifest = json.loads(out)
```
unused manifest can be replaced with "_"

Ethanlm replied:
I think keeping it a as a readable variable is easier to understand even though it is not used

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+          manifest = json.loads(out)
+          logging.debug("image tag exists for %s", image)
+        else:
+          manifest, manifest_hash = get_manifest_from_docker_image(pull_format, image)
```
unused manifest can be replaced with "_"

Ethanlm replied:
I think keeping it as a readable variable is easier to understand even though it is not used

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+
+  if bootstrap:
+    hdfs_dirs = [dest_root, dest_layers_dir, dest_config_dir, dest_manifest_dir]
+    image_tag_to_hash_path = hdfs_root + "/" + image_tag_to_hash
```
replace with args.hdfs_root or assign hdfs_root = args.hdfs_root like in some methods before.

Ethanlm replied:
This should be `dest_root`. Fixed and tested copy_update method.

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+      dest_layers_paths = [dest_layers_dir + "/" + layer + ".sqsh" for layer in src_layers]
+
+      logging.debug("Copying Manifest: %s", str(src_manifest_path))
+      logging.debug("Copying Layers: %s", str(src_layers_paths))
```
src_layers_paths is used here and on 1104, but is not defined/assigned anywhere.

Ethanlm replied:
missed 
```
src_layers_paths = [src_layers_dir + "/" + layer + ".sqsh" for layer in src_layers]
```

Fixed and tested copy_update

bipinprasad commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+        new_tags_and_comments = src_hash_to_tags.get(src_manifest_hash, None)
+        if new_tags_and_comments:
+          comment = ', '.join(map(str, new_tags_and_comments[1]))
+        if comment is None:
```
change this to "else:" or define comment = None before the line 1118 (otherwise comment can be undefined)

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/LocalOrHdfsImageTagToManifestPlugin.java:
```diff
@@ -0,0 +1,294 @@
+        return new BufferedReader(new InputStreamReader(fs.open(imageToHash)));
+    }
+
+    // You may specify multiple tags per hash all on the same line.
+    // Comments are allowed using #. Anything after this character will not
+    // be read
+    // Example file:
+    // foo/bar:current,fizz/gig:latest:123456789
+    // #this/line:wont,be:parsed:2378590895
+    //
+    // This will map both foo/bar:current and fizz/gig:latest to 123456789
```
Is this supposed to be part of javadoc on parsing logic?

Ethanlm replied:
will convert this to javadoc

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/LocalOrHdfsImageTagToManifestPlugin.java:
```diff
@@ -0,0 +1,294 @@
+
+            index = line.lastIndexOf(":");
+            if (index == -1) {
+                LOG.warn("Malformed imageTagToManifest entry: " + line);
```
It will be good to know which file or hdfs url this error is in. Same comment for line 214.

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/LocalOrHdfsImageTagToManifestPlugin.java:
```diff
@@ -0,0 +1,294 @@
+            input = fs.open(manifestPath);
+        } catch (IllegalArgumentException iae) {
+            throw new IOException("Manifest file is not a valid HDFS file: "
+                + manifestPath.toString(), iae);
```
The manifestPath seems to be a local file not HDFS.

Ethanlm replied:
This is on HDFS.
see
```
manifestDir = ObjectReader.getString(conf.get(DaemonConfig.STORM_OCI_IMAGE_HDFS_TOPLEVEL_DIR)) + "/manifests/";
```

bipinprasad commented on storm-client/src/jvm/org/apache/storm/daemon/supervisor/ClientSupervisorUtils.java:
```diff
@@ -61,10 +61,15 @@ static boolean doRequiredTopoFilesExist(Map<String, Object> conf, String stormId
 
     public static int processLauncherAndWait(Map<String, Object> conf, String user, List<String> args,
-                                             final Map<String, String> environment, final String logPreFix)
+                                             final Map<String, String> environment, final String logPreFix) throws IOException {
```
does this "final" do anything here?

Ethanlm replied:
I think it prevents the variable to be changed inside the method. Don't think it is very important so just left it as it is

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/HdfsManifestToResourcesPlugin.java:
```diff
@@ -0,0 +1,161 @@
+    public List<OciResource> getLayerResources(ImageManifest manifest) throws IOException {
+        List<OciResource> ociResources = new ArrayList<>();
+        for (ImageManifest.Blob blob : manifest.getLayers()) {
+            String mediaType = blob.getMediaType();
+            if (!mediaType.equals(LAYER_TAR_GZIP_MEDIA_TYPE)) {
+                throw new IOException("Invalid config mediaType: " + mediaType);
+            }
+
+            String[] layerDigest = blob.getDigest().split(":", 2);
+            String algorithm = layerDigest[0];
+            if (!algorithm.equals(LAYER_HASH_ALGORITHM)) {
+                throw new IOException("Invalid config digest algorithm: " + algorithm);
+            }
+
+            String hash = layerDigest[1];
+            if (!hash.matches(ALPHA_NUMERIC) || hash.length() != SHA256_HASH_LENGTH) {
+                throw new IOException("Malformed layer digest: " + hash);
+            }
+
```
Code duplicated in getConfigResource(). Should this be a method?

Ethanlm replied:
There are many variables involved. using one method is fine but might explore it as a follow-up

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/HdfsManifestToResourcesPlugin.java:
```diff
@@ -0,0 +1,161 @@
+    private FileSystem fs;
+    private LoadingCache<Path, FileStatus> statCache;
+
+    private static String CONFIG_MEDIA_TYPE = "application/vnd.docker.container.image.v1+json";
+
+    private static String LAYER_TAR_GZIP_MEDIA_TYPE = "application/vnd.docker.image.rootfs.diff.tar.gzip";
+
+    private static String SHA_256 = "sha256";
+
+    private static String CONFIG_HASH_ALGORITHM = SHA_256;
+
+    private static String LAYER_HASH_ALGORITHM = SHA_256;
+
+    private static int SHA256_HASH_LENGTH = 64;
+
+    private static String ALPHA_NUMERIC = "[a-zA-Z0-9]+";
```
These can be final

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/HdfsOciResourcesLocalizer.java:
```diff
@@ -0,0 +1,136 @@
+            // create working dir, copy file here, and set readable, then move to final location.
+            // this allows the operation to be atomic in case the supervisor dies.
+            File workingDir = new File(dst.getParent() + "/working");
+            workingDir.mkdir();
```
nit: return value is not checked for failure

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/HdfsOciResourcesLocalizer.java:
```diff
@@ -0,0 +1,136 @@
+            if (!setReadable) {
+                throw new IOException("Couldn't set " + workingDst + " to be world-readable");
+            }
+            workingDst.renameTo(dst);
```
nit: return value is not checked for failure

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/LocalOrHdfsImageTagToManifestPlugin.java:
```diff
@@ -0,0 +1,294 @@
+    private int ociCacheRefreshIntervalSecs;
+    private long lastRefreshTime;
+
+    private static String LOCAL_OR_HDFS_IMAGE_TAG_TO_MANIFEST_PLUGIN_PREFIX = "storm.oci.local.or.hdfs.image.tag.to.manifest.plugin.";
+
+    /**
+     * The HDFS location where the oci image-tag-to-hash file exists.
+     */
+    private static String HDFS_OCI_IMAGE_TAG_TO_HASH_FILE =
+        LOCAL_OR_HDFS_IMAGE_TAG_TO_MANIFEST_PLUGIN_PREFIX + "hdfs.hash.file";
+
+    /**
+     * The local file system location where the oci image-tag-to-hash file exists.
+     */
+    private static String LOCAL_OCI_IMAGE_TAG_TO_HASH_FILE =
+        LOCAL_OR_HDFS_IMAGE_TAG_TO_MANIFEST_PLUGIN_PREFIX + "local.hash.file";
+
+    /**
+     * The interval in seconds between refreshing the oci image-Tag-to-hash cache.
+     */
+    private static String OCI_CACHE_REFRESH_INTERVAL =
+        LOCAL_OR_HDFS_IMAGE_TAG_TO_MANIFEST_PLUGIN_PREFIX + "cache.refresh.interval.secs";
+
+    /**
+     * The number of manifests to cache.
+     */
+    private static String OCI_NUM_MANIFESTS_TO_CACHE = LOCAL_OR_HDFS_IMAGE_TAG_TO_MANIFEST_PLUGIN_PREFIX + "num.manifests.to.cache";
+
+    private static int SHA256_HASH_LENGTH = 64;
+
+    private static String ALPHA_NUMERIC = "[a-zA-Z0-9]+";
```
Can be final

bipinprasad commented on external/storm-hdfs-oci/src/main/java/org/apache/storm/container/oci/LocalOrHdfsImageTagToManifestPlugin.java:
```diff
@@ -0,0 +1,294 @@
+        }
+    }
+
+    private class LruCache extends LinkedHashMap<String, ImageManifest> {
```
Class can be static

bipinprasad commented on storm-core/src/native/worker-launcher/impl/configuration.c:
```diff
@@ -74,17 +74,17 @@ void free_configurations() {
   struct stat file_stat;
   if (stat(file, &file_stat) != 0) {
-    fprintf(ERRORFILE, "Can't stat file %s - %s\n", file, strerror(errno));
+    fprintf(ERRORFILE, "ERROR: Can't stat file %s - %s\n", file, strerror(errno));
```
nit: consistent indent size of 4 spaces (or two as used here mostly)
nit: stylewise, c code if expressions do not need a != 0 or != NULL part,
           i.e. "if (file_stat.st_uid)" instead of "if (file_stat.st_uid != 0)"

bipinprasad commented on storm-core/src/native/worker-launcher/impl/main.c:
```diff
@@ -269,6 +274,45 @@ int main(int argc, char **argv) {
       return INVALID_ARGUMENT_NUMBER;
     }
     exit_code = signal_container_as_user(user_detail->pw_name, container_pid, signal);
+  } else if (strcasecmp("run-oci-container", command) == 0) {
```
nit: why not the standard strcmp()?

Ethanlm replied:
I don't have strong opinion on making it case-sensitive. But we have been using case-insensitive in other commands here, don't think of changing it. case-insensitive is fine with me.

bipinprasad commented on storm-core/src/native/worker-launcher/impl/main.c:
```diff
@@ -269,6 +274,45 @@ int main(int argc, char **argv) {
+    if (argc != 6) {
+      fprintf(ERRORFILE, "Incorrect number of arguments (%d vs 6) for run-oci-container\n", argc);
+      fflush(ERRORFILE);
+      return INVALID_ARGUMENT_NUMBER;
```
nit: log and error files are closed before return for most cases. set exit_code here also?

Ethanlm replied:
I don't understand the question. Can you please elaborate?

bipinprasad replied:
Set the exit code to non-zero so that it is returned at the end of this function (where it properly closes open files). Although I don't think it is critical, since open files will be closed by the system on program exit.

Ethanlm replied:
got it. Thanks

bipinprasad commented on storm-core/src/native/worker-launcher/impl/oci/oci_launch_cmd.c:
```diff
@@ -0,0 +1,544 @@
+  return all_sections_ok;
+}
+
+static bool is_valid_oci_config_linux_seccomp(const cJSON* ocls) {
```
This JSON validation function is not implemented. Can it be similar to is_valid_oci_config_linux() ?

Ethanlm replied:
secomp file is a little more involved so leaving it for future.

bipinprasad commented on storm-core/src/native/worker-launcher/impl/oci/oci_write_config.c:
```diff
@@ -0,0 +1,464 @@
+    return NULL;
+  }
+
+  int rc = storm_user_info_fetch(sui, username);
```
nit: why is the method named storm_user_info(). Same question for storm_user_info.c and storm_user_info.h. They seem to be generic user_info query functions and nothing storm specific.

Ethanlm replied:
will make it generic

bipinprasad commented on storm-core/src/native/worker-launcher/impl/utils/file-utils.c:
```diff
@@ -0,0 +1,182 @@
+      }
+    }
+  }
+  if (bytes_left < 0) {
```
when is this supposed to happen?

Ethanlm replied:
I think this is not needed

bipinprasad commented on storm-core/src/native/worker-launcher/impl/utils/file-utils.c:
```diff
@@ -0,0 +1,182 @@
+  char *ret;
+  int bytesPrinted = asprintf(&ret, "%s/%s", dir, file);
+  if (bytesPrinted == -1) {
+    fprintf(ERRORFILE, "ERROR: failed to allocate file path\n");
```
Add info about exiting, like: "ERROR:EXITING process: failed to allocate....." because of the next exit() call.

bipinprasad commented on storm-core/src/native/worker-launcher/impl/utils/storm_user_info.c:
```diff
@@ -0,0 +1,275 @@
+    }
+    return 0;
+  } else if (ret != -1) {
+    // Any return code that is not -1 is considered as error.
```
nit: change to all "All negative return codes except -1 is considered as error"

Ethanlm replied:
when it is on linux, we have
```
if (ret > 0) {
} else if (ret != -1) {
    //ret could be zero
}
```
I will avoid causing confusion here. The original statement look fine to me.

bipinprasad commented on storm-core/src/native/worker-launcher/impl/utils/storm_user_info.c:
```diff
@@ -0,0 +1,275 @@
+  uinfo->gids = ngids;
+  uinfo->gids_size = ngroups;
+  ret = getgrouplist(uinfo->pwd.pw_name, uinfo->pwd.pw_gid, 
+                         uinfo->gids, &ngroups);
```
nit: Is line break formatting based on 60 character line? Should be 80-120 for readability

bipinprasad commented on storm-core/src/native/worker-launcher/impl/worker-launcher.c:
```diff
@@ -93,41 +109,41 @@ char* get_executable() {
   if (binary_gid != getgid()) {
     fprintf(LOGFILE, "ERROR: The configured worker-launcher group %d is different from"
-            " the group of the executable %d\n", getgid(), binary_gid);
+                     " the group of the executable %d\n",
```
Is this check needed here: if (filestat.st_mode & S_IGUID) == 0) ?

Ethanlm replied:
yes the worker-launcher binary needs to have setuid set.

bipinprasad commented on storm-server/src/main/java/org/apache/storm/container/oci/OciContainerExecutorConfig.java:
```diff
@@ -0,0 +1,1309 @@
+    public OciContainerExecutorConfig(String username,
+                                      String containerId,
+                                      String pidFile, String containerScriptPath, String containerCredentialsPath,
+                                      List<String> localDirs,
+                                      List<String> logDirs, List<OciLayer> layers, int reapLayerKeepCount,
```
containerCredentialsPath, localDirs and logDirs are unused vars.

bipinprasad commented on storm-server/src/main/java/org/apache/storm/container/oci/OciContainerExecutorConfig.java:
```diff
@@ -0,0 +1,1309 @@
+                this.options = options;
+            }
+
+            OciMount(String destination, String source, List<String> options) {
```
Unused constructor.

bipinprasad commented on storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java:
```diff
@@ -0,0 +1,631 @@
+    //CPU CFS (Completely Fair Scheduler) period
+    private static final long CPU_CFS_PERIOD_US = 100000;
+
+    private Map<String, Long> workerToContainerPid = new ConcurrentHashMap<>();
+    private Map<String, ExitCodeCallback> workerToExitCallback = new ConcurrentHashMap<>();
+    private Map<String, String> workerToUser = new ConcurrentHashMap<>();
```
Can these be final?

bipinprasad commented on storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java:
```diff
@@ -0,0 +1,631 @@
+
+        Long memoryInBytes = null;
+        if (workerToMemoryMb.containsKey(workerId)) {
+            memoryInBytes = workerToMemoryMb.get(workerId) * 1024 * 1024L;
```
Could this multiplication on the right side cause an overflow - int * int * long  when the first two integers are being multiplied.

Ethanlm replied:
nice catch.

bipinprasad commented:
This is a big commit and adds a very large set of functionalities.
 👍

agresch reviewed (commented):
review still in progress

agresch commented on storm-core/src/native/worker-launcher/impl/oci/oci_write_config.c:
```diff
@@ -0,0 +1,464 @@
+static cJSON* build_runc_config_process(const oci_launch_cmd* olc) {
+  cJSON* process = cJSON_CreateObject();
+  if (process == NULL) {
+    return NULL;
```
error messages?

agresch commented on storm-core/src/native/worker-launcher/impl/utils/cJSON.c:
```diff
@@ -0,0 +1,2932 @@
+/*
+  Copyright (c) 2009-2017 Dave Gamble and cJSON contributors
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy
```
do we need to adjust any licensing due to this?

Ethanlm replied:
I don't think so. Hadoop uses this the same way

agresch commented on storm-core/src/native/worker-launcher/impl/utils/file-utils.c:
```diff
@@ -0,0 +1,182 @@
+  uid_t user = geteuid();
+  gid_t group = getegid();
+  if (change_effective_user_to_wl() != 0) {
+    fputs("Cannot change to worker-launcher user\n", ERRORFILE);
```
prepend ERROR to messages here?

agresch commented on storm-core/src/native/worker-launcher/impl/utils/string-utils.h:
```diff
@@ -0,0 +1,98 @@
+ * Validate if the input is a valid container id
+ * return false/true
+ */
+bool validate_container_id(const char* input);
```
seems like this falls outside of string type utils?

Ethanlm replied:
Strictly, yes. Should I move it to `oci_launch_cmd.c` where it is used?

agresch commented on storm-server/src/main/java/org/apache/storm/container/oci/OciContainerManager.java:
```diff
@@ -118,6 +118,8 @@ public void reserveResourcesForWorker(String workerId, Integer workerMemoryMb, I
     public void releaseResourcesForWorker(String workerId) {
         workerToCpu.remove(workerId);
         workerToMemoryMb.remove(workerId);
+        workerToCores.remove(workerId);
```
if for some reason this PR is not committed, this looks like a separate PR that should be addressed to get in

agresch commented on storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java:
```diff
@@ -0,0 +1,631 @@
+                        checkContainersAlive();
+                    } catch (Exception e) {
+                        //Ignore
+                        LOG.debug("The CheckRuncContainerAlive thread has exception. Ignored", e);
```
Is it expected that checkContainersAlive() throw an exception without a message?

Ethanlm replied:
changing to warn

agresch commented on storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java:
```diff
@@ -0,0 +1,631 @@
+        if (pid != null) {
+            signal(pid, 9, user);
+        } else {
+            LOG.warn("Trying to forceKill container {} but pidfile is not found", workerId);
```
looks like it should be "container for workerId {}"

agresch commented on storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java:
```diff
@@ -0,0 +1,631 @@
+    @Override
+    public void releaseResourcesForWorker(String workerId) {
+        super.releaseResourcesForWorker(workerId);
+        workerToContainerPid.remove(workerId);
```
should we release workerToExitCallback?

Ethanlm replied:
yes it is done in `cleanup`.
Moving `workerToContainerPid.remove(workerId);` to `cleanup` as it makes more sense to me

agresch commented on storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java:
```diff
@@ -0,0 +1,631 @@
+        }
+        //remove from the watched list
+        LOG.debug("Removing {} from the watched workers list", workerId);
+        workerToUser.remove(workerId);
```
call releaseResourcesForWorker() instead?  

It looks like releaseResourcesForWorker() is called before cleanup() already.  

Do we need both?  

I didn't see a javadoc for cleanup().  I'm fuzzy on the difference.  I am all for combining into one if we can.

Ethanlm replied:
makes sense. keeping `cleanup` and removing `releaseResourcesForWorker`

agresch replied:
Thanks for addressing this.

agresch commented on storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java:
```diff
@@ -0,0 +1,631 @@
+
+        Long containerPid = getContainerPid(workerId);
+        if (containerPid == null) {
+            LOG.error("Counldn't get container PID for the worker {}. Skip profiling", workerId);
```
Counldn't spelling

Ethanlm commented on bin/docker-to-squash.py:
```diff
@@ -0,0 +1,1430 @@
+    hdfs_setrep(replication, hdfs_tmp_path)
+    hdfs_chmod("444", hdfs_tmp_path)
+
+    jar_path = HADOOP_PREFIX + "/share/hadoop/tools/lib/hadoop-extras-*.jar"
```
yes. HADOOP_PREFIX is where hadoop is installed. The rest is standard. 
It is looking for the hadoop-extras jar, e.g.:
```
share/hadoop/tools/lib/hadoop-extras-2.10.1.19.2106161351.jar
```

Ethanlm commented on .travis.yml:
```diff
@@ -51,6 +51,7 @@ before_install:
   - nvm install 8.9.3
   - nvm use 8.9.3
   - sudo apt-get update
+  - sudo apt-get install libssl-dev
```
oci.c needs `<openssl/evp.h>`

Ethanlm commented on docs/OCI-support.md:
````diff
@@ -0,0 +1,1704 @@
+}
+```
+
+Note: To use the `docker-to-squash.py`, you need to install [skopeo](https://github.com/containers/skopeo), [jq](https://stedolan.github.io/jq/) and squashfs-tools.
````
yes it is already there.

Ethanlm commented on docs/OCI-support.md:
````diff
@@ -0,0 +1,1704 @@
+}
+```
+
+Note: To use the `docker-to-squash.py`, you need to install [skopeo](https://github.com/containers/skopeo), [jq](https://stedolan.github.io/jq/) and squashfs-tools.
````
We are only referring to it and assuming they are available for use before running `docker-to-squash.py`. I think it should be fine.

Ethanlm commented on docs/OCI-support.md:
```diff
@@ -0,0 +1,1704 @@
+| `storm.oci.image.hdfs.toplevel.dir`      |  The HDFS location under which the oci image manifests, layers and configs directories exist.
+| `storm.oci.image.tag.to.manifest.plugin` |  The plugin to be used to get the image-tag to manifest mappings.
+| `storm.oci.localorhdfs.image.tag.to.manifest.plugin.hdfs.hash.file`   |   The hdfs location of image-tag to manifest mapping file. You need to set it if `org.apache.storm.container.oci.LocalOrHdfsImageTagToManifestPlugin` is used as `storm.oci.image.tag.to.manifest.plugin`.
+| `storm.oci.manifest.to.resources.plugin` | The plugin to be used to get oci resource according to the manifest.
```
`storm.oci.manifest.to.resources.plugin` and `storm.oci.resources.localizer` has no default value. We only have HDFS related plugins as of now, and we don't include `storm-hdfs-oci` in the package distribution by default

Ethanlm commented on storm-core/src/native/worker-launcher/Makefile.am:
```diff
@@ -16,14 +16,16 @@
 
 AM_CFLAGS=-I$(srcdir)/impl -Wall -g -Werror
 
+LIBS=-lcrypto
```
yes

Ethanlm commented on storm-core/src/native/worker-launcher/impl/main.c:
```diff
@@ -269,6 +274,45 @@ int main(int argc, char **argv) {
+    const char* command_file = argv[optind++];
+    const char* worker_artifacts_dir = argv[optind];
+    exit_code = setup_dir_permissions(working_dir, 1, TRUE);
+    if (exit_code == 0) {
```
it goes to the end of main.c and return the exit code.

Ethanlm commented on storm-core/src/native/worker-launcher/impl/main.c:
```diff
@@ -269,6 +274,45 @@ int main(int argc, char **argv) {
+    exit_code = setup_dir_permissions(working_dir, 1, TRUE);
+    if (exit_code == 0) {
+      exit_code = setup_worker_tmp_permissions(working_dir);
+      if (exit_code == 0) {
```
it goes to the end of main.c and return the exit code.

Ethanlm commented on storm-core/src/native/worker-launcher/impl/oci/oci.c:
```diff
@@ -0,0 +1,881 @@
+//Add this so it can compile on older version of linux kernel,
+//but certain runc related functionalities will not work during runtime.
+#ifndef LOOP_CTL_GET_FREE
+#define LOOP_CTL_GET_FREE  0x4C82
```
see https://man7.org/linux/man-pages/man4/loop.4.html
```
       LOOP_CTL_GET_FREE
              Allocate or find a free loop device for use.  On success,
              the device number is returned as the result of the call.
              This operation takes no argument.
```
Adding this definition, if not set already, is to make sure it can still compile on older version of linux kernel.

Ethanlm commented on storm-core/src/native/worker-launcher/impl/oci/oci_base_ctx.c:
```diff
@@ -0,0 +1,301 @@
+bool open_oci_base_ctx(oci_base_ctx* ctx) {
+  ctx->run_root = setup_oci_run_root_directories();
+  if (ctx->run_root == NULL) {
+    return false;
```
`setup_oci_run_root_directories` has sufficient log inside the method

Ethanlm commented on storm-core/src/native/worker-launcher/impl/oci/oci.c:
```diff
@@ -0,0 +1,881 @@
+
+static void destroy_oci_mount_ctx(oci_mount_ctx* ctx) {
+  if (ctx != NULL) {
+    free(ctx->src_path);
```
`free(null)` is okay.

Ethanlm commented on storm-core/src/native/worker-launcher/impl/oci/oci_launch_cmd.c:
```diff
@@ -0,0 +1,544 @@
+static void free_olc_layers(olc_layer_spec* layers, unsigned int num_layers) {
+  unsigned int i;
+  for (i = 0; i < num_layers; ++i) {
+    free(layers[i].media_type);
```
the size layers array matches with num_layers, so we don't need to check null here.

Ethanlm commented:
Addressed most of the comments, except 
1) add more comments in docker-to-squash.py.
2) add error messages in c code where it is missing
3) 2~3 unanswered questions about oci c code.

Will address these as follow-ups.

agresch reviewed (commented):
a couple minor issues and I am good with deferring my other comments.

Ethanlm commented:
Filed follow up jiras: 
STORM-3787 STORM-3788 STORM-3789
